File: Data Structures & Algorithms/reorder-linked-list/submission-3.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// this problem is essentially asking for us to interleave the linked list
// one from the front, one from the back, etc etc 

// that can be reframed into three primary parts
// need to split the array into two halves
// reverse the second half
// merge the two together

class Solution 
{
public:
    void reorderList(ListNode* head) 
    {
        if (!head->next) { return; }
        // use a fast & slow pointer to get the half way part
        // we'll start fast one above, which lands slow at the end of the first half
        // odd len linked lists will have slow longer by one which is fine since we
        // interleave with slow first
        ListNode* slow = head;
        ListNode* fast = head->next;

        while (fast && fast->next)
        {
            slow = slow->next;
            fast = fast->next->next;
        }

        // At the end of that slow is the last node in the first half
        ListNode* cur = slow->next;
        slow->next = nullptr; // this breaks the two LLs apart

        // Now we reverse our second half
        ListNode* prev = nullptr;
        while (cur)
        {
            ListNode* next = cur->next;
            cur->next = prev;
            prev = cur;
            cur = next;
        }

        // prev will now hold the new head of our reversed list
        cur = head;
        ListNode* cur2 = prev;
        while (cur2)
        {
            ListNode* next = cur->next;
            ListNode* next2 = cur2->next;

            cur->next = cur2;
            cur2->next = next;

            cur = next;
            cur2 = next2;
        }

    }
};

File: Data Structures & Algorithms/add-two-numbers/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        head = ListNode(-1)
        
        carry = 0
        cur = head

        while l1 and l2:
            new_val = l1.val + l2.val + carry
            cur.next = ListNode(new_val % 10)
            carry = new_val // 10

            cur = cur.next
            l1 = l1.next
            l2 = l2.next

        # Three scenarios here
        # 1. L1 is longer
        # 2. L2 is longer
        # 3. There is a leftover carry

        while l1:
            new_val = l1.val + carry
            cur.next = ListNode(new_val % 10)
            carry = new_val // 10
            l1 = l1.next
            cur = cur.next
        
        while l2:
            new_val = l2.val + carry
            cur.next = ListNode(new_val % 10)
            carry = new_val // 10
            l2 = l2.next
            cur = cur.next

        if carry > 0:
            cur.next = ListNode(1)

        return head.next

        
        